go/ql/lib/semmle/go/frameworks/RsCors.qll

```diff
@@ -0,0 +1,166 @@
+   */
+  class New extends Function {
+    New() { exists(Function f | f.hasQualifiedName(packagePath(), "New") | this = f) }
+  }
+
+  /**
+   * A write to the value of Access-Control-Allow-Credentials header
+   */
+  class AllowCredentialsWrite extends UniversalAllowCredentialsWrite {
+    DataFlow::Node base;
+
+    AllowCredentialsWrite() {
+      exists(Field f, Write w |
+        f.hasQualifiedName(packagePath(), "Options", "AllowCredentials") and
+        w.writesField(base, f, this) and
+        this.getType() instanceof BoolType
+      )
+    }
+
+    /**
+     * Get options struct holding header values
+     */
+    override DataFlow::Node getBase() { result = base }
+
+    /**
+     * Get options variable holding header values
+     */
+    override RsOptions getConfig() {
+      exists(RsOptions gc |
+        (
+          gc.getV().getBaseVariable().getDefinition().(SsaExplicitDefinition).getRhs() =
+            base.asInstruction() or
+          gc.getV().getAUse() = base
+        ) and
+        result = gc
+      )
+    }
+  }
+
+  /**
+   * A write to the value of Access-Control-Allow-Origins header
+   */
+  class AllowOriginsWrite extends UniversalOriginWrite {
+    DataFlow::Node base;
+
+    AllowOriginsWrite() {
+      exists(Field f, Write w |
+        f.hasQualifiedName(packagePath(), "Options", "AllowedOrigins") and
+        w.writesField(base, f, this) and
+        this.asExpr() instanceof SliceLit
+      )
+    }
+
+    /**
+     * Get options struct holding header values
+     */
+    override DataFlow::Node getBase() { result = base }
+
+    /**
+     * Get options variable holding header values
+     */
+    override RsOptions getConfig() {
+      exists(RsOptions gc |
+        (
+          gc.getV().getBaseVariable().getDefinition().(SsaExplicitDefinition).getRhs() =
+            base.asInstruction() or
+          gc.getV().getAUse() = base
+        ) and
+        result = gc
+      )
+    }
+  }
+
+  /**
+   * A write to the value of Access-Control-Allow-Origins of value "*", overriding AllowOrigins
+   */
+  class AllowAllOriginsWrite extends UniversalAllowAllOriginsWrite {
+    DataFlow::Node base;
+
+    AllowAllOriginsWrite() {
+      exists(Field f, Write w |
+        f.hasQualifiedName(packagePath(), "Options", "AllowAllOrigins") and
+        w.writesField(base, f, this) and
+        this.getType() instanceof BoolType
+      )
+    }
+
+    /**
+     * Get options struct holding header values
+     */
+    override DataFlow::Node getBase() { result = base }
+
+    /**
+     * Get options variable holding header values
+     */
+    override RsOptions getConfig() {
+      exists(RsOptions gc |
+        (
+          gc.getV().getBaseVariable().getDefinition().(SsaExplicitDefinition).getRhs() =
+            base.asInstruction() or
+          gc.getV().getAUse() = base
+        ) and
+        result = gc
+      )
+    }
+  }
+
+  /**
+   * A variable of type Options that holds the headers to be set.
```

Comment:
This isn't actually correct: if `v` is `a.b.c` then `this` is the `Variable a` and the thing that has type `Options` is `a.b.c`. I didn't notice this when you originally modeled GinCors. Can I ask why you are doing it this way? Is there a particular case where just using `Variable` or just using `SsaWithFields` doesn't work?

Reply:
Looking at it more carefully, I see that `RsOptions` and `GinConfig` are confined to be local variables defined in functions. Is that what you intended? Might you ever want to reason about one that is defined at file scope or package scope?

Reply:
@owen-mc 

> This isn't actually correct: if `v` is `a.b.c` then `this` is the `Variable a` and the thing that has type `Options` is `a.b.c`. I didn't notice this when you originally modeled GinCors. Can I ask why you are doing it this way? Is there a particular case where just using `Variable` or just using `SsaWithFields` doesn't work?

I can change this to just use SSAWithFields. I will ensure that the getBaseVariable() is removed. I see the problem with the current model is that since I am using SSAVariable, getSourceVariable's result is by definition a local variable. I would like to support package variables as well. I have tried to remove SSAs all together and just use DataFlow::Node, but cannot find a way for two DataFlow nodes to know if they actually represent the same variable. If you can think of any way to do this that would be great. If that doesn't work, is there any way to use SSAs that include package level variables? 

```go
var opts cors.Options                            <----- the package opts variables 

func rs_vulnerable() {
	opts.AllowedMethods = []string{"POST"}
	opts.AllowedOrigins = []string{"null"} <--- how to compare DataFlow::Node corresponding to this opts 
	opts.AllowCredentials = true              <---- to  DataFlow::Node corresponding to this opts and see that they both correspond to the package level opts
```

Reply:
Sorry I wasn't clearer about why the way they are modeled means they have to be local variables. Well done for figuring it out - coming back to this after four months, it took me a few minutes, even with your explanation.

I would say that SSAWithFields works great as long as you don't mind missing out writes to global variables. You may decide that in practice, that pattern doesn't come up and you're okay with not spotting it.

When dealing with just nodes, I think the normal way to think about them "representing the same variable" is whether there is value/taint flow from one to another.

There is an approach to this kind of thing in `go/ql/src/experimental/CWE-1004/AuthCookie.qll`, but I'm not sure it's necessarily one that I'd recommend.

I'll ask the rest of the codeql-go team if they can think of a better way of doing it.

Reply:
@owen-mc let me know if you have any updates on this :)

Reply:
Yes, sorry I didn't get back to you. The best way to include global variables is to copy [this code](https://github.com/github/codeql/blob/5e4a5c157174540547f48c1647f83a345a085412/go/ql/lib/semmle/go/dataflow/SSA.qll#L305-L402) which defines `SsaWithFields`and replace the root case with a read from a global variable. I guess `GlobalWithFields` would be the obvious name.

Reply:
@owen-mc tbh this seems to be a bit complex for my CodeQL skills, but I have written some initial code in the spirit of what you suggested Let me know if I'm on a the right the path (ignore any of the comments). I need [getDefinition](https://github.com/github/codeql/blob/f06ad7c7284d9e28093be5eba706b7f982cdca4b/go/ql/lib/semmle/go/dataflow/SsaImpl.qll#L288) but it uses [getLocalDefinition](https://github.com/github/codeql/blob/f06ad7c7284d9e28093be5eba706b7f982cdca4b/go/ql/lib/semmle/go/dataflow/SsaImpl.qll#L246) but I'm not too familiar with basic blocks in CodeQL to create a version for global. Any advice would be great.